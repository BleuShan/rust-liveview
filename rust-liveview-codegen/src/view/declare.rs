use crate::helpers::{
    Deref,
    IntoIterator,
};
use proc_macro::{
    Span,
    TokenStream,
};
use proc_macro2::TokenStream as TokenStream2;
use proc_macro_error::*;
use quote::quote;
use syn::{
    braced,
    parse::{
        Parse,
        ParseStream,
        Result,
    },
    punctuated::Punctuated,
    token::Brace,
    Expr,
    FieldValue,
    Ident,
    Member,
    Token,
};

#[derive(Debug)]
pub struct ElementDeclaration {
    ident: Ident,
    brace_token: Brace,
    fields: Punctuated<FieldValue, Token![,]>,
}

#[derive(IntoIterator, Deref, Debug)]
pub struct ElementDeclarations(Punctuated<ElementDeclaration, Token![,]>);

impl ElementDeclaration {
    fn generate_fields(&self) -> impl Iterator<Item = TokenStream2> {
        self.fields.clone().into_iter().map(|field| {
            let attrs = &field.attrs;
            let name = match &field.member {
                Member::Named(v) => v,
                Member::Unnamed(_) => {
                    abort!(Span::call_site(), "Anonymous members are not supported")
                }
            };

            let ty = match field.expr {
                Expr::Path(p) => p,
                _ => abort!(Span::call_site(), "Unsupported expression"),
            };
            quote! {
                #(#attrs)*
                #name: Option<#ty>,
            }
        })
    }
}

impl From<ElementDeclaration> for TokenStream2 {
    #[inline]
    fn from(declaration: ElementDeclaration) -> Self {
        let ident = &declaration.ident;
        let fields = declaration.generate_fields();
        let doc = format!(" {} element.", ident);
        quote! {
            #[derive(Debug, Element)]
            #[doc = #doc]
            pub struct #ident<C> {
                #(#fields)*
                _c: std::marker::PhantomData<C>
            }
        }
    }
}

impl From<ElementDeclaration> for TokenStream {
    #[inline]
    fn from(declaration: ElementDeclaration) -> Self {
        Self::from(TokenStream2::from(declaration))
    }
}
impl From<ElementDeclarations> for TokenStream2 {
    #[inline]
    fn from(items: ElementDeclarations) -> Self {
        #[allow(clippy::redundant_closure)]
        let streams = items.into_iter().map(|item| TokenStream2::from(item));
        quote! {
           #(#streams)*
        }
    }
}

impl From<ElementDeclarations> for TokenStream {
    #[inline]
    fn from(declaration: ElementDeclarations) -> Self {
        Self::from(TokenStream2::from(declaration))
    }
}

impl Parse for ElementDeclaration {
    fn parse(input: ParseStream<'_>) -> Result<Self> {
        let ident = input.parse()?;
        let content;
        let brace_token = braced!(content in input);
        let fields = content.parse_terminated(FieldValue::parse)?;
        Ok(Self {
            ident,
            brace_token,
            fields,
        })
    }
}

impl Parse for ElementDeclarations {
    fn parse(input: ParseStream<'_>) -> Result<Self> {
        let decls = input.parse_terminated(ElementDeclaration::parse)?;
        Ok(Self(decls))
    }
}
