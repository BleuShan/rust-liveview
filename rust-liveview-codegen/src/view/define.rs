use crate::helpers::{
    Deref,
    IntoIterator,
};
use proc_macro::TokenStream;
use proc_macro2::TokenStream as TokenStream2;
use proc_macro_error::*;
use quote::quote;
use std::iter;
use syn::{
    braced,
    parse::{
        Parse,
        ParseStream,
        Result,
    },
    punctuated::Punctuated,
    spanned::Spanned,
    Attribute,
    Expr,
    ExprPath,
    ExprType,
    Ident,
    Token,
    Type,
};

#[derive(Debug)]
pub enum ElementDefinition {
    Unit {
        attrs: Vec<Attribute>,
        ident: Ident,
    },
    Struct {
        attrs: Vec<Attribute>,
        ident: Ident,
        fields: Punctuated<ExprType, Token![,]>,
    },
}

thread_local! {
    static GLOBAL_ATTRS: Punctuated<ExprType, Token![,]> = {
        let def = syn::parse_str::<TokenStream2>(
            "GlobalAttributes {
                    accesskey: attributes::SpacedSet<String>,
                    class: attributes::SpacedSet<attributes::Class>, 
                    id: attributes::Id, 
                    lang: attributes::LanguageTag,
                    title: String
                }",
        )
        .and_then(syn::parse2::<ElementDefinition>)
        .expect_or_abort("Failed to parse global attributes definitions.");

        match def {
          ElementDefinition::Struct{fields, ..} => fields,
          ElementDefinition::Unit {ident, ..} =>
            abort!(ident.span(), "Invalid global attributes definitions.")
        }
    }
}

fn global_attributes<'a>() -> Box<dyn Iterator<Item = ExprType> + 'a> {
    box GLOBAL_ATTRS.with(|attrs| attrs.clone().into_iter())
}

#[derive(IntoIterator, Deref, Debug)]
pub struct ElementDefinitions(Punctuated<ElementDefinition, Token![,]>);
type ElementDefinitionField = (Vec<Attribute>, Ident, Box<Type>);

impl ElementDefinition {
    fn attrs(&self) -> &Vec<Attribute> {
        match self {
            Self::Unit { attrs, .. } => attrs,
            Self::Struct { attrs, .. } => attrs,
        }
    }

    fn ident(&self) -> &Ident {
        match self {
            Self::Unit { ident, .. } => ident,
            Self::Struct { ident, .. } => ident,
        }
    }

    fn own_fields<'a>(&'a self) -> Box<dyn Iterator<Item = ExprType> + 'a> {
        match self {
            Self::Struct { fields, .. } => box fields.clone().into_iter(),
            _ => box iter::empty::<ExprType>(),
        }
    }

    fn fields<'a>(&'a self) -> Box<dyn Iterator<Item = ElementDefinitionField> + 'a> {
        let iter = global_attributes().chain(self.own_fields());
        box iter.map(|field| {
            let (attrs, name) = match *field.expr {
                Expr::Path(ExprPath { path, qself, attrs })
                    if qself.is_none() && path.segments.len() == 1 =>
                {
                    (attrs, path.segments[0].ident.clone())
                }
                e => abort!(e.span(), "Invalid field name."),
            };
            let ty = field.ty;

            (attrs, name, ty)
        })
    }
    fn generate_struct_fields<'a>(&'a self) -> Box<dyn Iterator<Item = TokenStream2> + 'a> {
        box self.fields().map(|item| {
            let (attrs, name, ty) = item;
            quote! {
                #(#attrs)*
                #name: Option<#ty>,
            }
        })
    }

    fn generate_struct(&self) -> TokenStream2 {
        let attrs = self.attrs();
        let ident = self.ident();
        let fields = self.generate_struct_fields();
        let doc = format!(" {} element.", ident);
        quote! {
            #[doc = #doc]
            #[derive(Element, Clone, Default, Debug)]
            #(#attrs)*
            pub struct #ident<C> {
                #(#fields)*
                _c: std::marker::PhantomData<C>
            }
        }
    }
}

impl From<ElementDefinition> for TokenStream2 {
    #[inline]
    fn from(definition: ElementDefinition) -> Self {
        let type_definition = definition.generate_struct();
        quote! {
            #type_definition
        }
    }
}

impl From<ElementDefinition> for TokenStream {
    #[inline]
    fn from(definition: ElementDefinition) -> Self {
        Self::from(TokenStream2::from(definition))
    }
}
impl From<ElementDefinitions> for TokenStream2 {
    #[inline]
    fn from(items: ElementDefinitions) -> Self {
        #[allow(clippy::redundant_closure)]
        let streams = items.into_iter().map(|item| TokenStream2::from(item));
        quote! {
           #(#streams)*
        }
    }
}

impl From<ElementDefinitions> for TokenStream {
    #[inline]
    fn from(definition: ElementDefinitions) -> Self {
        Self::from(TokenStream2::from(definition))
    }
}

impl Parse for ElementDefinition {
    fn parse(input: ParseStream<'_>) -> Result<Self> {
        let attrs = input.call(Attribute::parse_outer)?;
        let ident = input.parse()?;
        if input.peek(Token![,]) {
            Ok(Self::Unit { attrs, ident })
        } else {
            let content;
            let _ = braced!(content in input);
            let fields = content.parse_terminated(ExprType::parse)?;
            Ok(Self::Struct {
                attrs,
                ident,
                fields,
            })
        }
    }
}

impl Parse for ElementDefinitions {
    fn parse(input: ParseStream<'_>) -> Result<Self> {
        let definitions = input.parse_terminated(ElementDefinition::parse)?;
        Ok(Self(definitions))
    }
}
