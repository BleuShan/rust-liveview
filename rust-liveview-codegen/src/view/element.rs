use crate::helpers::{
    DarlingResultExt,
    Inflector,
};
use darling::{
    ast::Data,
    FromDeriveInput,
    FromField,
};
use proc_macro::TokenStream;
use proc_macro2::TokenStream as TokenStream2;
use proc_macro_error::*;
use quote::quote;
use syn::{
    Attribute,
    DeriveInput,
    Ident,
    Type,
    Visibility,
};

#[derive(Debug, FromField)]
#[darling(attributes(element), forward_attrs)]
struct ElementField {
    ident: Option<Ident>,
    attrs: Vec<Attribute>,
    vis: Visibility,
    ty: Type,
    #[darling(default, rename = "skip")]
    skipped: bool,
}

impl ElementField {
    fn ident(&self) -> &Ident {
        self.ident
            .as_ref()
            .expect_or_abort("Anonymous fields are unsupported")
    }
    fn attribute_name(&self) -> String {
        self.ident().to_string().to_kebab_case().to_lowercase()
    }

    fn is_option(&self) -> bool {
        if let Type::Path(ref typepath) = self.ty {
            if typepath.qself.is_none() {
                let mut segments = typepath.path.segments.iter();
                return segments.any(|segment| segment.ident == "Option");
            }
        }
        false
    }

    fn generate_attribute_display_impl(&self) -> TokenStream2 {
        let name = self.attribute_name();
        let ident = self.ident();
        if self.is_option() {
            quote! {
                if let Some(ref #ident) = self.#ident {
                    write!(f, " {}=\"{}\"", #name, #ident)?;
                }
            }
        } else {
            quote! {
                write!(f, " {}=\"{}\"", #name, self.#ident)?;
            }
        }
    }
}

#[derive(Debug, FromDeriveInput)]
#[darling(
    attributes(element),
    supports(struct_named, struct_unit),
    forward_attrs
)]
pub struct Element {
    ident: Ident,
    attrs: Vec<Attribute>,
    vis: Visibility,
    data: Data<(), ElementField>,
    #[darling(default)]
    self_closing: bool,
}

impl From<DeriveInput> for Element {
    fn from(input: DeriveInput) -> Self {
        Self::from_derive_input(&input).expect_or_abort("An error occured while parsing the input.")
    }
}

impl Element {
    fn fields<'a>(&'a self) -> Box<dyn Iterator<Item = &ElementField> + 'a> {
        if let Some(fields) = self.data.as_ref().take_struct() {
            box fields.fields.into_iter().filter(|field| !field.skipped)
        } else {
            box std::iter::empty::<&ElementField>()
        }
    }

    fn generate_node_impl(&self) -> TokenStream2 {
        let ident = &self.ident;
        let node_name = ident.to_string().to_kebab_case().to_lowercase();
        quote! {
            impl Node for #ident {
                const NODE_NAME: &'static str = #node_name;
            }
        }
    }

    fn generate_element_impl(&self) -> TokenStream2 {
        let ident = &self.ident;
        let fields = self.fields();
        let field_names = fields.map(|field| field.attribute_name());
        quote! {
            impl Element for #ident {
                const ATTRIBUTE_NAMES: &'static [&'static str] = &[#(#field_names,)*];
            }
        }
    }

    fn generate_display_impl_closing_tag(&self) -> TokenStream2 {
        if self.self_closing {
            quote! {
                f.write_str("/>")
            }
        } else {
            quote! {
                write!(f, "</{}>", Self::NODE_NAME)
            }
        }
    }

    fn generate_display_impl_attributes(&self) -> TokenStream2 {
        let fields = self.fields();
        let attributes = fields.map(|field| field.generate_attribute_display_impl());
        if self.self_closing {
            quote!(#(#attributes)*)
        } else {
            quote! {
                #(#attributes)*
                f.write_str(">")?;
            }
        }
    }

    fn generate_display_impl(&self) -> TokenStream2 {
        let ident = &self.ident;
        let attributes = self.generate_display_impl_attributes();
        let closing_tag = self.generate_display_impl_closing_tag();
        quote! {
            impl std::fmt::Display for #ident {
                #[inline]
                fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
                    write!(f, "<{}", Self::NODE_NAME)?;
                    #attributes
                    #closing_tag
                }
            }
        }
    }

    pub fn generate(self) -> TokenStream {
        let impls = vec![
            self.generate_node_impl(),
            self.generate_element_impl(),
            self.generate_display_impl(),
        ];
        TokenStream::from(quote! {
           #(
                #[automatically_derived]
                #[allow(unused_qualifications)]
                #impls
           )*
        })
    }
}
