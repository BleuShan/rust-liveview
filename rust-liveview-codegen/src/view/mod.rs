mod declare;
mod element;

pub use declare::{
    ElementDeclaration,
    ElementDeclarations,
};
pub use element::Element;
