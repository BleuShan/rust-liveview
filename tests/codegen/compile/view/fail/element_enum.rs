use rust_liveview::view::Element;

fn main() {}

#[derive(Debug, Element)]
enum EnumElement {
    Test,
}
