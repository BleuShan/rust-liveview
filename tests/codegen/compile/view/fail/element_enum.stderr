error: Unsupported shape `enum`

error: An error occured while parsing the input.
 --> $DIR/element_enum.rs:5:17
  |
5 | #[derive(Debug, Element)]
  |                 ^^^^^^^
